patches/server/1052-Bytecode-Modification-Framework.patch

```diff
@@ -0,0 +1,1252 @@
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/VersionedClassloaderBytecodeModifier.java b/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/VersionedClassloaderBytecodeModifier.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..0889ef4c241d2d894713e7c02f8a5afbf9a068d2
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/VersionedClassloaderBytecodeModifier.java
+@@ -0,0 +1,68 @@
++package io.papermc.paper.plugin.entrypoint.classloader.bytecode;
++
++import io.papermc.paper.plugin.configuration.PluginMeta;
++import io.papermc.paper.plugin.entrypoint.classloader.ClassloaderBytecodeModifier;
++import io.papermc.paper.plugin.entrypoint.classloader.bytecode.rules.RewriteRule;
++import java.lang.invoke.MethodHandles;
++import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
++import org.checkerframework.checker.nullness.qual.Nullable;
++import org.objectweb.asm.ClassReader;
++import org.objectweb.asm.ClassVisitor;
++import org.objectweb.asm.ClassWriter;
++import org.objectweb.asm.MethodVisitor;
++import org.objectweb.asm.Opcodes;
++import org.objectweb.asm.Type;
++import org.objectweb.asm.commons.GeneratorAdapter;
++
++public abstract class VersionedClassloaderBytecodeModifier extends ClassVisitor implements ClassloaderBytecodeModifier, ClassContext {
++
++    private @MonotonicNonNull String className;
++    private @Nullable String superClassName;
++
++    protected VersionedClassloaderBytecodeModifier(final int api) {
++        super(api);
++    }
++
++    @Override
++    public void visit(final int version, final int access, final String name, final @Nullable String signature, final @Nullable String superName, final String @Nullable [] interfaces) {
++        super.visit(version, access, name, signature, superName, interfaces);
++        this.className = name;
++        this.superClassName = superName;
++    }
++
++    @Override
++    public final byte[] modify(final PluginMeta config, final byte[] bytecode) {
++        final ClassReader cr = new ClassReader(bytecode);
++        final ClassWriter cw = new ClassWriter(cr, 0);
++        this.cv = cw;
++
++        cr.accept(this, 0);
++        return cw.toByteArray();
++    }
++
++    @Override
++    public String className() {
++        return this.className;
++    }
++
++    @Override
++    public @Nullable String superClassName() {
++        return this.superClassName;
++    }
++
++    protected static Class<?> defineGeneratedDelegates(final Class<? extends VersionedClassloaderBytecodeModifier> versionedClass, final RewriteRule finalRule) {
++        final String className = versionedClass.getName() + "_GeneratedDelegates";
++        final ClassWriter writer = new ClassWriter(ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS);
++        writer.visit(Opcodes.V1_8, Opcodes.ACC_PUBLIC, className.replace('.', '/'), null, Type.getInternalName(Object.class), null);
```

Comment:
no asm expert, but why 1.8 and not 17?

Reply:
I just copied that from our `ASMEventExecutorGenerator` when I was still figuring out how to generate classes. We can change it if we want. Idk if it matters or not.

Reply:
idk if this would break if you use records or whatever